components/authz-service/server/v2/rules_property_test.go

```diff
@@ -0,0 +1,60 @@
+package v2_test
+
+import (
+	"context"
+	"reflect"
+	"testing"
+
+	"github.com/leanovate/gopter"
+	"github.com/leanovate/gopter/gen"
+	"github.com/leanovate/gopter/prop"
+
+	api "github.com/chef/automate/api/interservice/authz/v2"
+)
+
+func TestCreateRuleCreatesRulesInStorage(t *testing.T) {
+	ctx := context.Background()
+	cl, _, _, seed := setupRules(t)
+
+	conditionsGen := gen.StructPtr(reflect.TypeOf(&api.Condition{}), map[string]gopter.Gen{
+		"Type": gen.OneConstOf(
+			api.ProjectRuleConditionTypes_CHEF_SERVERS,
+			api.ProjectRuleConditionTypes_CHEF_ORGS,
+			api.ProjectRuleConditionTypes_CHEF_ENVIRONMENTS,
+			api.ProjectRuleConditionTypes_ROLES,
+			api.ProjectRuleConditionTypes_CHEF_TAGS,
+			api.ProjectRuleConditionTypes_POLICY_GROUP,
+			api.ProjectRuleConditionTypes_POLICY_NAME,
+		),
+		"Values": gen.SliceOf(gen.AnyString()),
+	})
+	createRuleReqGen := gen.Struct(reflect.TypeOf(&api.CreateRuleReq{}), map[string]gopter.Gen{
+		"Id":        gen.RegexMatch("^[a-z0-9-]{1,64}$"),
+		"Name":      gen.AnyString(),
+		"ProjectId": gen.Identifier(),
+		"Type": gen.OneConstOf(
+			api.ProjectRuleTypes_NODE,
+			api.ProjectRuleTypes_EVENT,
+		),
+		"Conditions": gen.SliceOf(conditionsGen),
+	})
+
+	params := gopter.DefaultTestParametersWithSeed(seed)
```

Comment:
what does `gopter` do for us?

Reply:
The [readme](https://github.com/leanovate/gopter#synopsis) isn't the best, but it's a property based tester. What it looks like we are using it for here is having it generate lots of different combinations of all the conditions possibilities (since we have a lot of different condition types) so we can test `CreateRule` against many possible different condition inputs without manually managing how they are generated.